encryption.go

```diff
@@ -0,0 +1,204 @@
+
+	if password == "" {
+		return nil, ErrIncorrectSecretLength
+	}
+	if salt == "" {
+		return nil, ErrIncorrectSaltLength
+	}
+
+	encPassword := []byte(password)
+	encSalt := []byte(salt)
+
+	return pbkdf2.Key(encPassword, encSalt, LIVEKIT_PBKDF_ITERATIONS, LIVEKIT_KEY_SIZE_BYTES, sha256.New), nil
+
+}
+
+func DeriveKeyFromBytes(secret []byte) ([]byte, error) {
+	return DeriveKeyFromBytesCustomSalt(secret, LIVEKIT_SDK_SALT)
+}
+
+func DeriveKeyFromBytesCustomSalt(secret []byte, salt string) ([]byte, error) {
+
+	info := make([]byte, LIVEKIT_HKDF_INFO_BYTES)
+	encSalt := []byte(salt)
+
+	if secret == nil {
+		return nil, ErrIncorrectSecretLength
+	}
+	if salt == "" {
+		return nil, ErrIncorrectSaltLength
+	}
+
+	hkdfReader := hkdf.New(sha256.New, secret, encSalt, info)
+
+	key := make([]byte, LIVEKIT_KEY_SIZE_BYTES)
+	_, err := io.ReadFull(hkdfReader, key)
+	if err != nil {
+		return nil, err
+	}
+
+	return key, nil
+
+}
+
+// Take audio sample (body of RTP) encrypted by LiveKit client SDK, extract IV and decrypt using provided key
+// Encrypted sample format based on livekit client sdk
+// ---------+-------------------------+---------+----
+// payload  |IV...(length = IV_LENGTH)|IV_LENGTH|KID|
+// ---------+-------------------------+---------+----
+// First byte of audio frame is not encrypted and only authenticated
+// payload - variable bytes
+// IV - variable bytes (equal to IV_LENGTH bytes)
+// IV_LENGTH - 1 byte
+// KID (Key ID) - 1 byte - ignored here, key is provided as parameter to function
+func DecryptGCMAudioSample(sample, key, sifTrailer []byte) ([]byte, error) {
+
+	if len(key) != 16 {
+		return nil, ErrIncorrectKeyLength
+	}
+
+	if sifTrailer != nil && len(sample) >= len(sifTrailer) {
+		possibleTrailer := sample[len(sample)-len(sifTrailer):]
+		if bytes.Equal(possibleTrailer, sifTrailer) {
+			// this is unencrypted Server Injected Frame (SIF) that should be dropped
+			return nil, nil
+		}
+
+	}
+
+	// variable naming is kept close to LiveKit client SDK decrypt function
+	// https://github.com/livekit/client-sdk-js/blob/main/src/e2ee/worker/FrameCryptor.ts#L402
+
+	frameHeader := sample[:unencrypted_audio_bytes] // first unencrypted bytes are "frameHeader" and used for authentication later
+	frameTrailer := sample[len(sample)-2:]          // last 2 bytes having IV_LENGTH and KID (1 byte each)
+	ivLength := int(frameTrailer[0])                // single byte, Endianness doesn't matter
+	ivStart := len(sample) - len(frameTrailer) - ivLength
+	if ivStart < 0 {
+		return nil, ErrIncorrectIVLength
+	}
+
+	iv := make([]byte, ivLength)
+	copy(iv, sample[ivStart:ivStart+ivLength]) // copy IV value out of sample into iv
+
+	cipherTextStart := len(frameHeader)
+	cipherTextLength := len(sample) - len(frameTrailer) - ivLength - len(frameHeader)
+	cipherText := make([]byte, cipherTextLength)
+	copy(cipherText, sample[cipherTextStart:cipherTextStart+cipherTextLength])
+
+	// setup AES
+	aesCipher, err := aes.NewCipher(key)
```

Comment:
Is there any performance penalty to creating this for every sample? Can this be created in track context and re-used?

Reply:
I've tried to keep it as separate and self-contained as possible without the need to instantiate anything separately to use it. I believe the AES operations are highly optimized in Go with the use of Go Assembler, but I'll see if I can run a quick benchmark with and without the cipher reuse.